Userland/Libraries/LibSemVer/SemVer.cpp

```diff
@@ -0,0 +1,305 @@
+/*
+ * Copyright (c) 2023, Gurkirat Singh <tbhaxor@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/CharacterTypes.h>
+#include <AK/Find.h>
+#include <AK/Function.h>
+#include <AK/GenericLexer.h>
+#include <AK/ReverseIterator.h>
+#include <AK/StringBuilder.h>
+#include <LibRegex/Regex.h>
+#include <LibRegex/RegexParser.h>
+#include <LibSemVer/SemVer.h>
+
+namespace SemVer {
+String SemVer::suffix() const
+{
+    StringBuilder sb;
+    if (m_prerelease_identifiers.size() > 0)
+        sb.appendff("-{}", prerelease());
+    if (m_buildmetadata.code_points().length() > 0)
+        sb.appendff("+{}", buildmetadata());
+    return MUST(sb.to_string());
+}
+
+String SemVer::to_string() const
+{
+    return MUST(String::formatted("{}{}{}{}{}{}", m_major, m_number_separator, m_minor, m_number_separator, m_patch, suffix()));
+}
+
+ErrorOr<SemVer> SemVer::bump(BumpType type) const
+{
+    switch (type) {
+    case BumpType::Major:
+        return SemVer(m_major + 1, 0, 0, m_number_separator);
+    case BumpType::Minor:
+        return SemVer(m_major, m_minor + 1, 0, m_number_separator);
+    case BumpType::Patch:
+        return SemVer(m_major, m_minor, m_patch + 1, m_number_separator);
+    case BumpType::Prerelease: {
+        Vector<String> prerelease_identifiers = m_prerelease_identifiers;
+        bool is_found = false;
+
+        // Unlike comparision, prerelease bumps take from RTL
+        for (auto& identifier : AK::ReverseWrapper::in_reverse(prerelease_identifiers)) {
+            auto numeric_identifier = identifier.to_number<u32>();
+            if (numeric_identifier.has_value()) {
+                is_found = true;
+                identifier = TRY(String::formatted("{}", numeric_identifier.value() + 1));
+                break;
+            }
+        }
+
+        // Append 0 identifier if there is no numeric found to be bumped
+        if (!is_found)
+            prerelease_identifiers.append("0"_string);
+
+        return SemVer(m_major, m_minor, m_patch, m_number_separator, prerelease_identifiers, ""_string);
+    }
+    default:
+        return Error::from_string_view("Invalid bump type"sv);
+    }
+}
+
+bool SemVer::is_same(SemVer const& other, CompareType compare_type) const
+{
+    switch (compare_type) {
+    case CompareType::Major:
+        return m_major == other.m_major;
+    case CompareType::Minor:
+        return m_major == other.m_major && m_minor == other.m_minor;
+    case CompareType::Patch:
+        return m_major == other.m_major && m_minor == other.m_minor && m_patch == other.m_patch;
+    default:
+        return to_string() == other.to_string();
+    }
+}
+
+bool SemVer::is_greater_than(SemVer const& other) const
+{
+    // Priortize the normal version string
+    // Precedence is determined by the first difference when comparing them from left to right
```

Comment:
These sentences are missing dots at the end (same as basically all other sentence-looking comments in here).

Reply:
Makes sense. But I have left some comments as it is because they contain version string or links.